### ISSUE TITLE
AD: change the backend transparently

### ISSUE THREAD

**Issue by jbcaillau** [control-toolbox/CTBase.jl#25]
- we currently use [ForwardDiff](https://github.com/JuliaDiff/ForwardDiff.jl) but should be able to move transparently (= without any user code change) to another backend
- [x] use the backend choice as a default / global behaviour, not *hard coded*
- [x] as soon as sth more efficient than dual numbers ([Enzyme](https://enzyme.mit.edu/julia), *e.g.*) is available, switch

**Comment by jbcaillau** [comment]
It is an issue. Now in `CTBase.jl`, FWIW

**Comment by ocots** [comment]
I am not convinced that this is an issue. It is more a wish :-)

**Comment by jbcaillau** [comment]
We should move from `ForwardDiff` to [AbstractDifferentiation](https://github.com/JuliaDiff/AbstractDifferentiation.jl)

**Comment by jbcaillau** [comment]
See also: [FastDifferentiation.jl](https://brianguenter.github.io/FastDifferentiation.jl/dev/)

**Comment by ocots** [comment]
Check also [DifferentiationInterface.jl](https://github.com/gdalle/DifferentiationInterface.jl).

**Comment by gdalle** [comment]
Friendly ping from the creator of DifferentiationInterface: I'm available to help you make the transition if you want me to :)

**Comment by ocots** [comment]
Hi! This would be great. I propose first to post here how we use AD. @jbcaillau and @PierreMartinon, please complete.

First, we have define some auxiliary functions: 

https://github.com/control-toolbox/CTBase.jl/blob/b73f56d6447e1967bf56c59f49dc8bcdb09dd71c/src/utils.jl#L76

https://github.com/control-toolbox/CTBase.jl/blob/b73f56d6447e1967bf56c59f49dc8bcdb09dd71c/src/utils.jl#L101

**Comment by gdalle** [comment]
Thanks for the links, I'll take a look but I already have a few questions.

Why do you call the derivative the gradient?

Are the derivative and Jacobian the only operators you need? What are the typical input and output dimensionalities for the Jacobian? Depending on the answer, you may want to parametrize with different AD backends for the derivative (forward mode always) and the Jacobian (forward mode for large input and small output, reverse mode for small input and large output, otherwise unclear).

Do you take derivatives or Jacobians of the same function several times, but with different input vectors? If so, you will hugely benefit from a preparation mechanism like the one that is implemented in DifferentiationInterface.

**Comment by gdalle** [comment]
As for ADNLPModels, they are also considering a switch to DifferentiationInterface but it might be slightly slower

**Comment by jbcaillau** [comment]
@gdalle 
 
> Why do you call the derivative the gradient? What is this ctNumber that you use?

`ctNumber = Real`. We want to deal more or less uniformly with reals and one dimensional vectors, that is why the special case when the variable is a single real is explicitly dealt with.
 
> Are the derivative and Jacobian the only operators you need? What are the typical input and output dimensionalities for the Jacobian? Depending on the answer, you may want to parametrize with different AD backends for the derivative (forward mode always) and the Jacobian (forward mode for large input and small output, reverse mode for small input and large output, otherwise unclear).

Dimensions < 1e2, *e.g.* to build the right hand side of a Hamiltonian system.

> Do you take derivatives or Jacobians of the same function several times, but with different input vectors? If so, you will hugely benefit from a preparation mechanism like the one that is implemented in DifferentiationInterface.

✅ to be tested elsewhere (see also [this comment](https://github.com/control-toolbox/CTBase.jl/pull/140#issuecomment-2138315017))

**Comment by ocots** [comment]
I think that for `CTBase.jl`, a step has been done. Do we close this issue? We will see next in `CTFlows.jl` how to handle this.

**Comment by gdalle** [comment]
To me this is not yet done, because #141 added a `backend` kwargs to `ctgradient` and the like, but this kwarg is not passed from further up the chain. As a result, users cannot change the AD backend, even though package developers can through the `__auto()` function.

**Comment by jbcaillau** [comment]
@gdalle check this [PR](https://github.com/control-toolbox/CTBase.jl/pull/160)

**Comment by gdalle** [comment]
To clarify, even with this PR, you're currently doing something like

```julia
function solve_control_problem(f)
    # ...
    for i in 1:n
        x -= gradient(f, x)
    end
    # ...
end

function gradient(f, x, backend=default_backend())
    # ...
end
```

And for users who only care about high-level interfaces, and who never call `gradient` directly, the following seems better to me:

```julia
function solve_control_problem(f, backend)
    # ...
    for i in 1:n
        x -= gradient(f, x, backend)
    end
    # ...
end

function gradient(f, x, backend=default_backend())
    # ...
end
```

But you know best if that's relevant in your case or not.

**Comment by ocots** [comment]
We totally agree with you. The second choice is better.

But, actually the function to solve optimal control problems is not in the `CTBase.jl` package. 

Besides, our function 

```julia
function gradient(f, x, backend=default_backend())
    # ...
end
```

is not used in the resolution function of optimal control problems. It is used for instance in the package `CTFlows.jl` [here](https://github.com/control-toolbox/CTFlows.jl/blob/8526bf24f289513775dd5a06f92a12d9ec9ab3c1/ext/hamiltonian.jl#L46). I agree that here I will have to add a kwarg for the AD backend.

About the resolution of the optimal control problems, we pass through `ADNLPModels.jl` and again we want the user to have the possibility to choose the AD backend.

**Comment by jbcaillau** [comment]
@gdalle agreed, thanks for the feedback. actually, there is now a setter that allows user / dev to change the backend (globally and dynamically); it is also easy to add optional kwarg to allow this anywhere it makes sense (solvers, *etc.*) We leave this issue open for further testing, *e.g.* for cases requiring a change of backend between first order derivative computation and second order ones.

On a side note: check this upcoming [talk](https://pretalx.com/juliacon2024/talk/AADQNR) at JuliaCon 2024 (we'll also be [around](https://pretalx.com/juliacon2024/talk/AADQNR))

**Comment by gdalle** [comment]
Thanks for pointing out ADOLC.jl, we're already on the ball ;) see https://github.com/TimSiebert1/ADOLC.jl/issues/7 to track progress

**Comment by ocots** [comment]
No more AD here. AD is only in CTFlows now.
